java-checks/src/main/java/org/sonar/java/checks/helpers/ClassPattternsUtils.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * SonarQube Java
+ * Copyright (C) 2012-2023 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks.helpers;
+
+import java.util.List;
+import org.sonar.java.model.ModifiersUtils;
+import org.sonar.plugins.java.api.tree.ClassTree;
+import org.sonar.plugins.java.api.tree.MethodTree;
+import org.sonar.plugins.java.api.tree.Modifier;
+import org.sonar.plugins.java.api.tree.ModifiersTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.VariableTree;
+
+public class ClassPattternsUtils {
+
+  private ClassPattternsUtils() {
+  }
+
+  public static boolean isPrivateInnerClass(ClassTree classTree) {
+    return !classTree.symbol().owner().isPackageSymbol() &&
+      ModifiersUtils.hasModifier(classTree.modifiers(), Modifier.PRIVATE);
+  }
+
+  public static boolean isUtilityClass(ClassTree classTree) {
+    return hasOnlyStaticMembers(classTree) && !anonymousClass(classTree) && !extendsAnotherClassOrImplementsSerializable(classTree)
```

Comment:
We can probably swap the first 2 tests since the first does not need to iterate through all the members
```suggestion
    return !anonymousClass(classTree) && hasOnlyStaticMembers(classTree) && !extendsAnotherClassOrImplementsSerializable(classTree)
```